ext/src/Cassandra/DefaultSchema.c

```diff
@@ -0,0 +1,152 @@
+
+  if (meta == NULL) {
+    RETURN_NULL();
+  }
+
+  object_init_ex(return_value, cassandra_default_keyspace_ce);
+  keyspace = (cassandra_keyspace*) zend_object_store_get_object(return_value TSRMLS_CC);
+  keyspace->schema = php_cassandra_add_ref(self->schema);
+  keyspace->meta   = meta;
+}
+
+PHP_METHOD(DefaultSchema, keyspaces)
+{
+  cassandra_schema* self;
+  CassIterator* iterator;
+  const CassSchemaMeta* meta;
+  const CassSchemaMetaField* field;
+  const CassValue* value;
+  const char* keyspace_name;
+  size_t keyspace_name_len;
+  zval* zend_keyspace;
+  cassandra_keyspace* keyspace;
+
+  if (zend_parse_parameters_none() == FAILURE)
+    return;
+
+  self     = (cassandra_schema*) zend_object_store_get_object(getThis() TSRMLS_CC);
+  iterator = cass_iterator_from_schema((CassSchema*) self->schema->data);
+
+  array_init(return_value);
+  while (cass_iterator_next(iterator)) {
+    meta  = cass_iterator_get_schema_meta(iterator);
+    field = cass_schema_meta_get_field(meta, "keyspace_name");
+    value = cass_schema_meta_field_value(field);
+
+    ASSERT_SUCCESS_BLOCK(cass_value_get_string(value, &keyspace_name, &keyspace_name_len),
+      zval_ptr_dtor(&return_value);
+      return;
+    );
+
+    MAKE_STD_ZVAL(zend_keyspace);
+    object_init_ex(zend_keyspace, cassandra_default_keyspace_ce);
+    keyspace = (cassandra_keyspace*) zend_object_store_get_object(zend_keyspace TSRMLS_CC);
+    keyspace->schema = php_cassandra_add_ref(self->schema);
+    keyspace->meta   = meta;
+    add_assoc_zval_ex(return_value, keyspace_name, keyspace_name_len + 1, zend_keyspace);
+  }
+
+  cass_iterator_free(iterator);
+}
+
+ZEND_BEGIN_ARG_INFO_EX(arginfo_name, 0, ZEND_RETURN_VALUE, 1)
+  ZEND_ARG_INFO(0, name)
+ZEND_END_ARG_INFO()
+
+ZEND_BEGIN_ARG_INFO_EX(arginfo_none, 0, ZEND_RETURN_VALUE, 0)
+ZEND_END_ARG_INFO()
+
+static zend_function_entry cassandra_default_schema_methods[] = {
+  PHP_ME(DefaultSchema, keyspace, arginfo_name, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultSchema, keyspaces, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_FE_END
+};
+
+static zend_object_handlers cassandra_default_schema_handlers;
+
+static HashTable*
+php_cassandra_default_schema_properties(zval *object TSRMLS_DC)
+{
+  /* cassandra_schema* self = (cassandra_schema*) zend_object_store_get_object(object TSRMLS_CC); */
+  HashTable* props = zend_std_get_properties(object TSRMLS_CC);
+
+  return props;
+}
+
+static int
+php_cassandra_default_schema_compare(zval *obj1, zval *obj2 TSRMLS_DC)
+{
+  if (Z_OBJCE_P(obj1) != Z_OBJCE_P(obj2))
+    return 1; /* different classes */
+
+  return Z_OBJ_HANDLE_P(obj1) != Z_OBJ_HANDLE_P(obj1);
+}
+
+static void
+php_cassandra_default_schema_free(void *object TSRMLS_DC)
+{
+  cassandra_schema* schema = (cassandra_schema*) object;
+
+  zend_object_std_dtor(&schema->zval TSRMLS_CC);
+
+  if (schema->schema) {
+    php_cassandra_del_ref(&schema->schema);
+  }
+
+  efree(schema);
```

Comment:
yes, something like the following:

```php
<?php

$keyspace = $session->schema()->keyspace("abc");
```

the Cassandra\Keyspace object would have a reference to CassSchema, while the Cassandra\Schema object would get GC'd by falling out of scope, since we don't preserve the result of `schema()` call anywhere.

Did I understand your question correctly?